renderdoc/driver/metal/metal_command_buffer.cpp

```diff
@@ -0,0 +1,122 @@
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ ******************************************************************************/
+
+#include "metal_command_buffer.h"
+#include "core/settings.h"
+#include "metal_device.h"
+#include "metal_resources.h"
+
+WrappedMTLCommandBuffer::WrappedMTLCommandBuffer(MTL::CommandBuffer *realMTLCommandBuffer,
+                                                 ResourceId objId, WrappedMTLDevice *wrappedMTLDevice)
+    : WrappedMTLObject(realMTLCommandBuffer, objId, wrappedMTLDevice, wrappedMTLDevice->GetStateRef())
+{
+  objcBridge = AllocateObjCBridge(this);
+}
+
+template <typename SerialiserType>
+bool WrappedMTLCommandBuffer::Serialise_presentDrawable(SerialiserType &ser, MTL::Drawable *drawable)
+{
+  SERIALISE_ELEMENT_LOCAL(CommandBuffer, this);
+
+  SERIALISE_CHECK_READ_ERRORS();
+
+  // TODO: implement RD MTL replay
+  if(IsReplayingAndReading())
+  {
+  }
+  return true;
+}
+
+void WrappedMTLCommandBuffer::presentDrawable(MTL::Drawable *drawable)
+{
+  SERIALISE_TIME_CALL(Unwrap(this)->presentDrawable(drawable));
+  if(IsCaptureMode(m_State))
+  {
+    Chunk *chunk = NULL;
+    {
+      CACHE_THREAD_SERIALISER();
+      SCOPED_SERIALISE_CHUNK(MetalChunk::MTLCommandBuffer_presentDrawable);
+      Serialise_presentDrawable(ser, drawable);
+      chunk = scope.Get();
+    }
+    MetalResourceRecord *record = GetRecord(this);
+    record->AddChunk(chunk);
+    record->cmdInfo->present = true;
+    record->cmdInfo->drawable = drawable;
+  }
+  else
+  {
+    // TODO: implement RD MTL replay
+  }
+}
+
+template <typename SerialiserType>
+bool WrappedMTLCommandBuffer::Serialise_commit(SerialiserType &ser)
+{
+  SERIALISE_ELEMENT_LOCAL(CommandBuffer, this);
+
+  SERIALISE_CHECK_READ_ERRORS();
+
+  // TODO: implement RD MTL replay
+  if(IsReplayingAndReading())
+  {
+    CommandBuffer->commit();
+  }
+  return true;
+}
+
+void WrappedMTLCommandBuffer::commit()
+{
+  SERIALISE_TIME_CALL(Unwrap(this)->commit());
+  if(IsCaptureMode(m_State))
+  {
+    Chunk *chunk = NULL;
+    {
+      CACHE_THREAD_SERIALISER();
+      SCOPED_SERIALISE_CHUNK(MetalChunk::MTLCommandBuffer_commit);
+      Serialise_commit(ser);
+      chunk = scope.Get();
+    }
+    GetResourceManager()->MarkResourceFrameReferenced(GetResID(m_WrappedMTLCommandQueue),
```

Comment:
The marking of the parent wrapped command queue might change in the future.
See [here](https://github.com/Zorro666/renderdoc/issues/58) for the thoughts and goals about the chunks for the creation of Permanent Resources.

Reply:
In a future PR, this is likely to be changed to only mark the command queue as referenced during an active capture.
```
bool capframe = IsActiveCapturing(m_State);
if(capframe)
{
    GetResourceManager()->MarkResourceFrameReferenced(GetResID(m_WrappedMTLCommandQueue), eFrameRef_Read);
}
```

Reply:
I'm not sure this mark reference here is necessary or the best fit for what I'd expect. If the queue is the parent of a command buffer then its creation record will be included regardless as long as the command buffer is marked as referenced. Generally speaking marking a resource referenced should happen when it is explicitly used, so e.g. if a queue operation happened mid-frame then that would mark the queue as referenced. It would feel more natural to mark the command buffer itself as referenced since this function is the one which submits the work in it, though since command buffers usually have special lifetime management it's maybe more likely that neither are needed.

Also yes, resources should usually only be marked as referenced during active capturing.